src/ompl/tools/experience/src/ExperienceSetup.cpp

```diff
@@ -0,0 +1,162 @@
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Dave Coleman
+   Desc:   Common functions for experience-based planning. Used by lighting and thunder
+*/
+
+#include "ompl/tools/experience/ExperienceSetup.h"
+#include "ompl/tools/multiplan/ParallelPlan.h"
+
+// Boost
+#include <boost/filesystem.hpp>
+
+ompl::tools::ExperienceSetup::ExperienceSetup(const base::SpaceInformationPtr &si)
+  : ompl::geometric::SimpleSetup(si)
+{
+  logInitialize();
+};
+
+ompl::tools::ExperienceSetup::ExperienceSetup(const base::StateSpacePtr &space)
+  : ompl::geometric::SimpleSetup(space)
+{
+  logInitialize();
+};
+
+void ompl::tools::ExperienceSetup::logInitialize()
+{
+  // Header of CSV file
+  csvDataLogStream_
+    // Times
+    << "planning_time,insertion_time,"
+    // Solution properties
+    << "planner,result,is_saved,"
+    // Failure booleans
+    << "approximate,too_short,insertion_failed,"
+    // Lightning properties
+    << "score,"
+    // Thunder (SPARS) properties
+    << "num_vertices,num_edges,num_connected_components,"
+    // Hack for using python cause im lazy right now
+    << "total_experiences,total_scratch,total_recall,total_failed,total_approximate,"
+    << "total_too_short,total_insertion_failed,"
+    << "avg_planning_time,avg_insertion_time" 
+    << std::endl;
+}
+
+void ompl::tools::ExperienceSetup::convertLogToString(const ExperienceLog &log)
+{
+  csvDataLogStream_ 
+    << log.planning_time << ","
+    << log.insertion_time << ","
+    << log.planner << ","
+    << log.result << ","
+    << log.is_saved << ","
+    << log.approximate << ","
+    << log.too_short << ","
+    << log.insertion_failed << ","
+    << log.score << ","
+    << log.num_vertices << ","
+    << log.num_edges << ","
+    << log.num_connected_components
+    //<< ",0,0,0,0,0,0,0,0,0"
+    << std::endl;
+}
+
+void ompl::tools::ExperienceSetup::saveDataLog(std::ostream &out)
+{
+    // Export to file and clear the stream
+    out << csvDataLogStream_.str();
+    csvDataLogStream_.str("");
+}
+
+bool ompl::tools::ExperienceSetup::getFilePath(const std::string &databaseName, const std::string &databaseDirectory)
+{
+    namespace fs = boost::filesystem;
+
+    // Check that the directory exists, if not, create it
```

Comment:
How about we move the burden to the caller? We assume we get a full (correct path) and don't do any env var checking. We could probably get rid of this function. We could even just deal with streams directly (it is up to the caller to figure out how to get the streams set up)